server/src/main/java/com/linecorp/centraldogma/server/internal/storage/project/ProjectApiManager.java

```diff
@@ -116,14 +129,34 @@ public CompletableFuture<Revision> unremoveProject(String projectName, Author au
         checkInternalProject(projectName, "unremove");
         // Restore the project first then update its metadata as 'active'.
         return commandExecutor.execute(Command.unremoveProject(author, projectName))
-                .thenCompose(unused -> metadataService.restoreProject(author, projectName));
+                              .thenCompose(unused -> metadataService.restoreProject(author, projectName));
     }
 
     public Project getProject(String projectName) {
-        if (isInternalProject(projectName) && !isAdmin()) {
+        return getProject(projectName, AuthUtil.currentUserOrNull());
+    }
+
+    public Project getProject(String projectName, @Nullable User user) {
+        final Project project = projectManager.get(projectName);
+
+        if (!isInternalProject(projectName)) {
+            return project;
+        }
+
+        if (user == null) {
             throw new IllegalArgumentException("Cannot access " + projectName);
         }
-        return projectManager.get(projectName);
+        if (user.isAdmin()) {
+            return project;
+        }
+        final ProjectMetadata metadata = project.metadata();
+        if (metadata != null) {
+            // Only show internal projects to the members of the project.
+            if (metadata.memberOrDefault(user.id(), null) != null) {
```

Comment:
For the record, do you mind explaining your reasoning on why you chose to check this here instead of at `RequiresPermissionDecorator`? Permission related logic is scattered across the code base, and I would like to organize in my mind which checks go where.

Reply:
Not all methods are decorated with `RequiresPermissionDecorator`. 
https://github.com/line/centraldogma/blob/c615ba5d4f0216b33e35f5141da54c068fb8b743/server/src/main/java/com/linecorp/centraldogma/server/internal/api/RepositoryServiceV1.java#L82-L84

It depends on `ApiRequestConverter` to get `Project`.
https://github.com/line/centraldogma/blob/53885c45e5ddf4e56eefa7162ea5ca743e711e85/server/src/main/java/com/linecorp/centraldogma/server/internal/api/converter/HttpApiRequestConverter.java#L61-L62

 I agree with your opinion about using `RequiresPermissionDecorator`. Currently `RequiresPermissionDecorator` is used for checking the permission for repositories but not for the members of projects. We may need some refactoring to use `RequiresPermissionDecorator` for project permission. As we will continue to update permissions, I will keep this as a point of improvement.

Reply:
Please understand that I cannot make changes to the PR right away because I need to deploy this feature to the internal cluster.